### PR TITLE
Fix GridLayoutManager Layout calculation

### LINKED ISSUE #7681: [GridLayoutManager] `ArrangeChildren` was affected by measure contraint
### Description

GridLayoutManager internally has `GridStructure` and it used for `ArrangeChildren`
`GridStructure` use contraint that passed on constructor. it make unreasonable result

for example
``` xml
                <Grid 
                    Margin="5"
                    HeightRequest="120"
                    WidthRequest="105">
                    <BoxView 
                        Color="{Binding Color}" />
                    <Label 
                        Text="{Binding Name}"
                        HorizontalOptions="Center"
                        VerticalOptions="Center" />
                </Grid>
```
On this view, 
If I measue with 300 x Inf
and then the result of measure is `{Width=105 Height=120}`

So, I arrange it to somewhere by ArrangeChildren with `{X=0 Y=0 Width=104.84641638225256 Height=120.13651877133105}`
I expect BoxView width is 105 but actually it was 300 because I measued GridLayout with 300
And also I expect BoxView height is 120 but actually it was 40 because I measued GridLayout with inf

I think, When a grid was arranged with a specific size, Children of grid should be arranged with that size

### Steps to Reproduce

1. Create Grid view like below 
``` xml
                <Grid 
                    Margin="5"
                    HeightRequest="120"
                    WidthRequest="105">
                    <BoxView 
                        Color="{Binding Color}" />
                    <Label 
                        Text="{Binding Name}"
                        HorizontalOptions="Center"
                        VerticalOptions="Center" />
                </Grid>
```

2. Measure it with some specific size 
3. Arrange it with measured size

### Version with bug

6.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, Android, Windows, macOS, Other (Tizen, Linux, etc. not supported by Microsoft directly)

### Affected platform versions

Not related with Platform versions

### Did you find any workaround?

https://github.com/dotnet/maui/blob/3ae0029bf0c0e44b52a6b9beaf9295f763f55966/src/Core/src/Layouts/GridLayoutManager.cs#L31

Always create a new _gridStructure with arranged size

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by myroot**
### Description of Change

 Replaced widthConstraint/heightConstraint to explicit width/height, if explicit width/height was valid.

 GridlayoutManager returns the explicit width/height as a measure result if it was valid
 https://github.com/dotnet/maui/blob/3ae0029bf0c0e44b52a6b9beaf9295f763f55966/src/Core/src/Layouts/GridLayoutManager.cs#L289
 https://github.com/dotnet/maui/blob/3ae0029bf0c0e44b52a6b9beaf9295f763f55966/src/Core/src/Layouts/GridLayoutManager.cs#L306

 But, Layout calculation was not affected by explicit width/height. so, `ArrangeChildren` is not working propertly
 It was fixed

### Issues Fixed

Fixes #7681

### CODE REVIEW

File: src/Core/src/Layouts/GridLayoutManager.cs [hunk 1]
```diff
@@ -84,11 +84,12 @@ public GridStructure(IGridLayout grid, double widthConstraint, double heightCons
 			{
 				_grid = grid;
 
-				_gridWidthConstraint = widthConstraint;
-				_gridHeightConstraint = heightConstraint;
-
 				_explicitGridHeight = _grid.Height;
 				_explicitGridWidth = _grid.Width;
+
+				_gridWidthConstraint = _explicitGridWidth > -1 ? _explicitGridWidth : widthConstraint;
```
[the remark this answers is not in the document]
Comment (myroot) [reply src/Core/src/Layouts/GridLayoutManager.cs:90]: Sure, but I not familiar with UnitTest structure so it takes a little time

[the remark this answers is not in the document]
Comment (myroot) [reply src/Core/src/Layouts/GridLayoutManager.cs:90]: @jsuarezruiz 
Hi, I'm done. could you review the testcase?

**Review by hartez — APPROVED** [review]

**Pull request merged**